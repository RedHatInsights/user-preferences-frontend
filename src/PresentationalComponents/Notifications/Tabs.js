import React, { useEffect, useRef, useState } from 'react';
import PropTypes from 'prop-types';
import { useFormApi } from '@data-driven-forms/react-form-renderer';
import { Text } from '@patternfly/react-core/dist/dynamic/components/Text';
import { Title } from '@patternfly/react-core/dist/dynamic/components/Title';
import { useLocation, useNavigate } from 'react-router-dom';
import { useFormState } from 'react-final-form';
import { getNavFromURL, setNavToURL } from './urlSync';
import TabsMenu from './TabsMenu';

const renderPageHeading = (bundleTitle, sectionTitle) => (
  <React.Fragment>
    <Title headingLevel="h3" size="xl" className="pf-v5-u-pb-xs">
      {`${sectionTitle} | ${bundleTitle}`}
    </Title>
    <Text className="pf-v5-u-mb-md">
      Configure your {sectionTitle} notifications.
    </Text>
  </React.Fragment>
);

const FormTabs = ({ fields, titleRef, bundles }) => {
  const location = useLocation();
  const navigate = useNavigate();
  const formOptions = useFormApi();
  const searchRef = useRef(null);
  const navConfig = useRef({});

  const [search, setSearch] = useState('');
  const [filteredFields, setFilteredFields] = useState(fields);
  const { pristine, submitSucceeded, dirtyFieldsSinceLastSubmit } =
    useFormState();

  const handleResize = () => {
    const container = document.getElementById('notifications-container');
    const gridElement = document.getElementById('notifications-grid');
    const menu = document.getElementById('notifications-menu-content');
    if (menu?.style && gridElement?.style) {
      const buttonsHeight =
        document
          .getElementById('user-pref__form-buttons')
          ?.getBoundingClientRect()?.height || 0;
      gridElement.style.height = `${
        container.getBoundingClientRect().height - buttonsHeight
      }px`;
      let menuMaxHeight;
      menuMaxHeight =
        container.getBoundingClientRect().height -
        titleRef.current.getBoundingClientRect().height -
        searchRef.current.getBoundingClientRect().height -
        (menuMaxHeight < menu.scrollHeight ? buttonsHeight : 0) -
        1;
      menu.style.maxHeight = `${menuMaxHeight}px`;
      menu.style.height = `${menuMaxHeight - buttonsHeight}px`;
    }
  };

  useEffect(() => {
    if (window.screen.width > 768) {
      window.addEventListener('resize', handleResize);
      handleResize();
    }

    navConfig.current = getNavFromURL(location, navigate, fields, {
      bundle: fields?.[0]?.name,
      app: fields?.[0]?.fields?.[0]?.name,
    });

    return () => window.removeEventListener('resize', handleResize);
  }, []);

  useEffect(() => {
    const filtered = fields.reduce(
      (acc, bundle) => [
        ...acc,
        {
          ...bundle,
          fields: bundle.fields.filter((item) => {
            const searchValue = search.toLowerCase();
            return (
              item.name?.toLowerCase().includes(searchValue) ||
              item.title?.toLowerCase().includes(searchValue)
            );
          }, []),
        },
      ],
      []
    );
    setFilteredFields(filtered);
  }, [search]);

  useEffect(() => {
    handleResize();
  }, [pristine, submitSucceeded, dirtyFieldsSinceLastSubmit]);

  return (
    <React.Fragment>
      <div className="pref-notifications--nav">
        <TabsMenu
          searchRef={searchRef}
          search={search}
          setSearch={setSearch}
          fields={filteredFields}
          onClick={(e, bundleName, sectionName) => {
            e.preventDefault();
            navConfig.current = {
              bundle: bundleName,
              app: sectionName,
            };
            setNavToURL(location, navigate, navConfig.current);
          }}
        />
      </div>
      <div className="pref-notifications--inputs">
        <React.Fragment>
          {renderPageHeading(
            bundles[navConfig.current.bundle]?.label,
            fields
              .reduce((acc, curr) => [...acc, ...curr.fields], [])
              .filter(
                (item) =>
                  item.name === navConfig.current.app &&
                  item.bundle === navConfig.current.bundle
              )?.[0]?.label
          )}
          <div>
            {formOptions.renderForm(
              fields.reduce(
                (acc, curr) => [
                  ...acc,
                  ...curr.fields.map((item) => ({
                    ...item,
                    key: `form-${item.bundle}-${item.name}`,
                    fields: [
                      item.fields.map((input) => ({
                        ...input,
                        hideField: !(
                          item.name === navConfig.current.app &&
                          item.bundle === navConfig.current.bundle
                        ),
                      })),
                    ],
                  })),
                ],
                []
              ),
              formOptions
            )}
          </div>
        </React.Fragment>
      </div>
    </React.Fragment>
  );
};

FormTabs.propTypes = {
  fields: PropTypes.array.isRequired,
  titleRef: PropTypes.any,
  bundles: PropTypes.shape({ label: PropTypes.string }),
};

export default React.memo(FormTabs);
