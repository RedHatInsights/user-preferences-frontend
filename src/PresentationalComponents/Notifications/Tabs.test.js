import React from 'react';
import { Form, RendererContext } from '@data-driven-forms/react-form-renderer';
import { fireEvent, getByText, render } from '@testing-library/react';
import Tabs from './Tabs';
import { Router } from 'react-router-dom';

describe('Tabs tests', () => {
  const setSearch = jest.fn();
  const replace = jest.fn();

  const mockHistory = {
    push: jest.fn(),
    replace,
    listen: jest.fn(),
    location: {},
  };

  const fields = [
    {
      title: 'Red Hat Enterprise Linux',
      name: 'rhel',
      fields: [
        {
          name: 'advisor',
          label: 'Advisor',
          component: 'tabGroup',
          fields: [],
          bundle: 'rhel',
        },
        {
          name: 'advisor2',
          label: 'Advisor2',
          component: 'tabGroup',
          fields: [{}],
          bundle: 'rhel',
        },
      ],
    },
  ];

  afterEach(() => {
    setSearch.mockReset();
    replace.mockReset();
  });

  it('should render correctly', () => {
    const { container } = render(
      <Router history={mockHistory}>
        <Form onSubmit={() => undefined}>
          {() => (
            <RendererContext.Provider
              value={{
                formOptions: {
                  renderForm: () => {},
                },
              }}
            >
              <Tabs
                search=""
                fields={fields}
                isLoading={false}
                bundles={{ rhel: { label: 'RHEL' } }}
              />
            </RendererContext.Provider>
          )}
        </Form>
      </Router>
    );
    expect(container).toMatchSnapshot();
  });

  it('should render loading correctly', () => {
    const { container } = render(
      <Router history={mockHistory}>
        <Form onSubmit={() => undefined}>
          {() => (
            <RendererContext.Provider
              value={{
                formOptions: {
                  renderForm: () => {},
                },
              }}
            >
              <Tabs search="" fields={[]} bundles={{}} isLoading />
            </RendererContext.Provider>
          )}
        </Form>
      </Router>
    );
    expect(container).toMatchSnapshot();
  });

  it('should replace history on click', () => {
    const { container } = render(
      <Router history={mockHistory}>
        <Form onSubmit={() => undefined}>
          {() => (
            <RendererContext.Provider
              value={{
                formOptions: {
                  renderForm: () => {},
                },
              }}
            >
              <Tabs search="" fields={fields} isLoading={false} bundles={{}} />
            </RendererContext.Provider>
          )}
        </Form>
      </Router>
    );
    expect(replace).toBeCalledTimes(1);
    fireEvent.click(getByText(container, 'Advisor2'));
    expect(replace).toHaveBeenNthCalledWith(2, {
      pathname: undefined,
      search: 'bundle=rhel&app=advisor2',
    });
  });
});
