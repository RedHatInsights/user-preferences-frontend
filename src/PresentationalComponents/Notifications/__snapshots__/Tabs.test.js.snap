// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tabs tests should render correctly 1`] = `
<div>
  <div
    class="pref-notifications--nav"
  >
    <div
      class="pf-v6-c-menu pf-m-plain pf-m-scrollable"
      data-ouia-component-id="OUIA-Generated-Menu-1"
      data-ouia-component-type="PF6/Menu"
      data-ouia-safe="true"
    >
      <div
        class="pf-v6-c-menu__search"
      >
        <div
          class="pf-v6-c-menu__search-input pf-v6-u-mx-sm"
        >
          <div
            class="pf-v6-c-text-input-group"
            customicon="[object Object]"
            data-testid="search-input"
            type="search"
          >
            <div
              class="pf-v6-c-text-input-group__main pf-m-icon"
            >
              <span
                class="pf-v6-c-text-input-group__text"
              >
                <span
                  class="pf-v6-c-text-input-group__icon"
                >
                  <svg
                    aria-hidden="true"
                    class="pf-v6-svg"
                    fill="currentColor"
                    height="1em"
                    role="img"
                    viewBox="0 0 512 512"
                    width="1em"
                  >
                    <path
                      d="M505 442.7L405.3 343c-4.5-4.5-10.6-7-17-7H372c27.6-35.3 44-79.7 44-128C416 93.1 322.9 0 208 0S0 93.1 0 208s93.1 208 208 208c48.3 0 92.7-16.4 128-44v16.3c0 6.4 2.5 12.5 7 17l99.7 99.7c9.4 9.4 24.6 9.4 33.9 0l28.3-28.3c9.4-9.4 9.4-24.6.1-34zM208 336c-70.7 0-128-57.2-128-128 0-70.7 57.2-128 128-128 70.7 0 128 57.2 128 128 0 70.7-57.2 128-128 128z"
                    />
                  </svg>
                </span>
                <input
                  aria-label="Filter menu items"
                  class="pf-v6-c-text-input-group__text-input"
                  placeholder="Search services"
                  type="text"
                  value=""
                />
              </span>
            </div>
          </div>
        </div>
      </div>
      <hr
        class="pf-v6-c-divider"
      />
      <div
        class="pf-v6-c-menu__content"
        id="notifications-menu-content"
      >
        <section
          class="pf-v6-c-menu__group pf-v6-u-px-sm pf-v6-u-font-size-xs"
        >
          <h1
            class="pf-v6-c-menu__group-title"
          >
            Red Hat Enterprise Linux
          </h1>
          <ul
            class="pf-v6-c-menu__list"
            role="menu"
          >
            <li
              class="pf-v6-c-menu__list-item pf-v6-u-font-size-md"
              role="none"
            >
              <button
                class="pf-v6-c-menu__item pf-v6-u-font-size-md"
                role="menuitem"
                tabindex="0"
                type="button"
              >
                <span
                  class="pf-v6-c-menu__item-main"
                >
                  <span
                    class="pf-v6-c-menu__item-text"
                  >
                    Advisor
                  </span>
                </span>
              </button>
            </li>
            <li
              class="pf-v6-c-menu__list-item pf-v6-u-font-size-md"
              role="none"
            >
              <button
                class="pf-v6-c-menu__item pf-v6-u-font-size-md"
                role="menuitem"
                tabindex="-1"
                type="button"
              >
                <span
                  class="pf-v6-c-menu__item-main"
                >
                  <span
                    class="pf-v6-c-menu__item-text"
                  >
                    Advisor2
                  </span>
                </span>
              </button>
            </li>
          </ul>
        </section>
      </div>
    </div>
  </div>
  <div
    class="pref-notifications--inputs"
  >
    <h3
      class="pf-v6-c-title pf-m-xl pf-v6-u-pb-xs"
      data-ouia-component-id="OUIA-Generated-Title-1"
      data-ouia-component-type="PF6/Title"
      data-ouia-safe="true"
    >
      Advisor | RHEL
    </h3>
    <p
      class="pf-v6-c-content--p pf-v6-u-mb-md"
      data-ouia-component-id="OUIA-Generated-Content-1"
      data-ouia-component-type="PF6/Content"
      data-ouia-safe="true"
      data-pf-content="true"
    >
      Configure your 
      Advisor
       notifications.
    </p>
    <div />
  </div>
</div>
`;

exports[`Tabs tests should render loading correctly 1`] = `
<div>
  <div
    class="pref-notifications--nav"
  >
    <div
      class="pf-v6-c-menu pf-m-plain pf-m-scrollable"
      data-ouia-component-id="OUIA-Generated-Menu-2"
      data-ouia-component-type="PF6/Menu"
      data-ouia-safe="true"
    >
      <div
        class="pf-v6-c-menu__search"
      >
        <div
          class="pf-v6-c-menu__search-input pf-v6-u-mx-sm"
        >
          <div
            class="pf-v6-c-text-input-group"
            customicon="[object Object]"
            data-testid="search-input"
            type="search"
          >
            <div
              class="pf-v6-c-text-input-group__main pf-m-icon"
            >
              <span
                class="pf-v6-c-text-input-group__text"
              >
                <span
                  class="pf-v6-c-text-input-group__icon"
                >
                  <svg
                    aria-hidden="true"
                    class="pf-v6-svg"
                    fill="currentColor"
                    height="1em"
                    role="img"
                    viewBox="0 0 512 512"
                    width="1em"
                  >
                    <path
                      d="M505 442.7L405.3 343c-4.5-4.5-10.6-7-17-7H372c27.6-35.3 44-79.7 44-128C416 93.1 322.9 0 208 0S0 93.1 0 208s93.1 208 208 208c48.3 0 92.7-16.4 128-44v16.3c0 6.4 2.5 12.5 7 17l99.7 99.7c9.4 9.4 24.6 9.4 33.9 0l28.3-28.3c9.4-9.4 9.4-24.6.1-34zM208 336c-70.7 0-128-57.2-128-128 0-70.7 57.2-128 128-128 70.7 0 128 57.2 128 128 0 70.7-57.2 128-128 128z"
                    />
                  </svg>
                </span>
                <input
                  aria-label="Filter menu items"
                  class="pf-v6-c-text-input-group__text-input"
                  placeholder="Search services"
                  type="text"
                  value=""
                />
              </span>
            </div>
          </div>
        </div>
      </div>
      <hr
        class="pf-v6-c-divider"
      />
      <div
        class="pf-v6-c-menu__content"
        id="notifications-menu-content"
      >
        <div
          class="pf-v6-c-empty-state pf-m-sm pf-v6-u-mt-lg"
        >
          <div
            class="pf-v6-c-empty-state__content"
          >
            <div
              class="pf-v6-c-empty-state__header"
            >
              <div
                class="pf-v6-c-empty-state__icon"
              >
                <svg
                  aria-hidden="true"
                  class="pf-v6-svg"
                  fill="currentColor"
                  height="1em"
                  role="img"
                  viewBox="0 0 512 512"
                  width="1em"
                >
                  <path
                    d="M505 442.7L405.3 343c-4.5-4.5-10.6-7-17-7H372c27.6-35.3 44-79.7 44-128C416 93.1 322.9 0 208 0S0 93.1 0 208s93.1 208 208 208c48.3 0 92.7-16.4 128-44v16.3c0 6.4 2.5 12.5 7 17l99.7 99.7c9.4 9.4 24.6 9.4 33.9 0l28.3-28.3c9.4-9.4 9.4-24.6.1-34zM208 336c-70.7 0-128-57.2-128-128 0-70.7 57.2-128 128-128 70.7 0 128 57.2 128 128 0 70.7-57.2 128-128 128z"
                  />
                </svg>
              </div>
              <div
                class="pf-v6-c-empty-state__title"
              >
                <h4
                  class="pf-v6-c-empty-state__title-text"
                >
                  No matching services found
                </h4>
              </div>
            </div>
            <div
              class="pf-v6-c-empty-state__body"
            >
              Adjust your filters and try again.
            </div>
            <div
              class="pf-v6-c-empty-state__footer"
            >
              <button
                class="pf-v6-c-button pf-m-link"
                data-ouia-component-id="OUIA-Generated-Button-link-1"
                data-ouia-component-type="PF6/Button"
                data-ouia-safe="true"
                type="button"
              >
                <span
                  class="pf-v6-c-button__text"
                >
                  Clear filters
                </span>
              </button>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
  <div
    class="pref-notifications--inputs"
  >
    <h3
      class="pf-v6-c-title pf-m-xl pf-v6-u-pb-xs"
      data-ouia-component-id="OUIA-Generated-Title-2"
      data-ouia-component-type="PF6/Title"
      data-ouia-safe="true"
    >
      undefined | undefined
    </h3>
    <p
      class="pf-v6-c-content--p pf-v6-u-mb-md"
      data-ouia-component-id="OUIA-Generated-Content-2"
      data-ouia-component-type="PF6/Content"
      data-ouia-safe="true"
      data-pf-content="true"
    >
      Configure your 
       notifications.
    </p>
    <div />
  </div>
</div>
`;
